feat: Allow functions to override the Mappers

blackspherefollower opened the pull request:
This change allows fromIdMappers and toIdMappers to be set to functions
instead of maps of functions. This allows complex state objects to be
stored as readble URL parameters with dynamic names.

For example:
```
state = {
  filters[
    { type: Brand, value: foo },
    { type: Device, value: var }
  ]
}
```
Renders as: http://localhost:8000/?filter0Brand=foo&filter1Device=bar

```
const reactUrlStateOptions = {
  fromIdResolvers: async (param, value, oldState) => {
    let newState = { ...oldState }
    let found = param.match(/filter(\d+)([A-Za-z].+)/)
    if (found !== null) {
      let fId = parseInt(found[1], 10)
      if (newState.filters === undefined) {
        newState.filters = new Array(fId + 1)
      } else {
        while (newState.filters.length < fId + 1) {
          newState.filters.push({})
        }
      }
      newState.filters[fId] = { type: found[2], value: value }
    }
    return newState
  },
  toIdMappers: (param, state) => {
    if (param === `filters`) {
      return state.filters
        .map((f, i) => {
          if (typeof f.toUrl === `function`) {
            return `filter${i}${f.type}=${f.value)}`
          }
          return undefined
        })
        .filter(p => p !== undefined && p !== null)
        .join(`&`)
    }
  },
  pathname: `/`,
}
```

blackspherefollower committed:
feat: Allow functions to override the Mappers

This change allows fromIdMappers and toIdMappers to be set to functions
instead of maps of functions. This allows complex state objects to be
stored as readble URL parameters with dynamic names.

For example:
```
state = {
  filters[
    { type: Brand, value: foo },
    { type: Device, value: var }
  ]
}
```
Renders as: http://localhost:8000/?filter0Brand=foo&filter1Device=bar

```
const reactUrlStateOptions = {
  fromIdResolvers: async (param, value, oldState) => {
    let newState = { ...oldState }
    let found = param.match(/filter(\d+)([A-Za-z].+)/)
    if (found !== null) {
      let fId = parseInt(found[1], 10)
      if (newState.filters === undefined) {
        newState.filters = new Array(fId + 1)
      } else {
        while (newState.filters.length < fId + 1) {
          newState.filters.push({})
        }
      }
      newState.filters[fId] = { type: found[2], value: value }
    }
    return newState
  },
  toIdMappers: (param, state) => {
    if (param === `filters`) {
      return state.filters
        .map((f, i) => {
          if (typeof f.toUrl === `function`) {
            return `filter${i}${f.type}=${f.value)}`
          }
          return undefined
        })
        .filter(p => p !== undefined && p !== null)
        .join(`&`)
    }
  },
  pathname: `/`,
}
```

blackspherefollower committed:
bug: Swapping history modes for server-side builds

If there's no dom (like in webpack builds), this the hisory call
throws errors.

blackspherefollower committed:
feat: Pass callback though initializeReactUrlState

This provides a means for a caller to add logic for default states
or other behaviours based on the URL parsing.

```
this.reactUrlState = initializeReactUrlState(this)(
  reactUrlStateOptions,
    () => {
      if (this.state.filters.length === 0) {
        this.setState({
          filters: [{ field: `Availability`, urlData: `Available,DIY` }],
      })
    }
  }
)
```

ZippoLag commented:
Hey @blackspherefollower just letting you know you have merge conflicts on this, you may get a better chance to have this merged if you solve them.

blackspherefollower commented:
@ZippoLag Thanks: hadn't been paying attention to the ongoing development.

blackspherefollower committed:
feat: Add option to prevent state from being set on init

This is useful in cases where the state from the URL parameters
will impact the state of the loading component and so would
cause additional state changes
